tests/partitions_test.py

```diff
@@ -163,41 +165,44 @@ def test_new_joiner_helps_liveness(network, args):
 
     primary, backups = network.find_nodes()
 
-    # Add a new node, but partition them before trusting them
-    new_node = network.create_node("local://localhost")
-    network.join_node(new_node, args.package, args, from_snapshot=False)
-    new_joiner_partition = [new_node]
-    new_joiner_rules = network.partitioner.partition(
-        [primary, *backups], new_joiner_partition
-    )
-
-    # Trust the new node, and wait for commit of this (but don't ask the new node itself, which doesn't know this yet)
-    network.trust_node(new_node, args, no_wait=True)
-    check_can_progress(primary)
-
-    # Partition the primary, temporarily creating a minority service that cannot make progress
-    minority_partition = backups[len(backups) // 2 :] + new_joiner_partition
-    minority_rules = network.partitioner.partition(minority_partition)
-    time.sleep(
-        network.observed_election_duration
-    )  # Ensure this node has become primary
-    with backups[0].client("user0") as c:
-        r = c.post("/app/log/private", {"id": 42, "msg": "Hello world"})
-        assert r.status_code == http.HTTPStatus.SERVICE_UNAVAILABLE
-
-    # Restore the new node to the service
-    new_joiner_rules.drop()
-
-    # Confirm that the new node catches up, and progress can be made in this majority partition
-    network.wait_for_new_primary(primary, minority_partition)
-    check_can_progress(new_node)
+    with contextlib.ExitStack() as stack:
+        # Add a new node, but partition them before trusting them
+        new_node = network.create_node("local://localhost")
+        network.join_node(new_node, args.package, args, from_snapshot=False)
+        new_joiner_partition = [new_node]
+        new_joiner_rules = stack.enter_context(
+            network.partitioner.partition([primary, *backups], new_joiner_partition)
+        )
 
-    # Explicitly drop rules before continuing
-    minority_rules.drop()
+        # Trust the new node, and wait for commit of this (but don't ask the new node itself, which doesn't know this yet)
+        network.trust_node(new_node, args, no_wait=True)
+        check_can_progress(primary)
 
-    network.wait_for_primary_unanimity()
-    primary, _ = network.find_nodes()
-    network.wait_for_all_nodes_to_commit(primary=primary)
+        # Partition the primary, temporarily creating a minority service that cannot make progress
+        minority_partition = backups[len(backups) // 2 :] + new_joiner_partition
+        minority_rules = stack.enter_context(
+            network.partitioner.partition(minority_partition)
+        )
+        time.sleep(
```

Comment:
Please replace this with a periodic poll timing out at network.observed_election_duration, in many cases we probably don't need to (and therefore want to) wait this long. The comment is also inaccurate it seems? Nothing checks for a new primary afaict.

Reply:
Added a comment explaining why this is a bit weird, and the sleep is the easiest solution.